### PR TITLE
support for execution environments in multi execution services

### PR DESCRIPTION
**Pull request by gs-gunjan**
#### What does this PR do / why is it needed ?

support for execution environments in multi execution services
support of services with only query lambda
execution environments are packageables elements comprising of mapping and runtime

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

No

### CODE REVIEW

File: legend-engine-pure-ide-light/src/main/resources/ideLightConfig.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
     "connector": {
       "maxRequestHeaderSize": "32KiB",
       "type": "http",
-      "port": 9200
```
Comment (gs-gunjan) [inline legend-engine-pure-ide-light/src/main/resources/ideLightConfig.json:24]: will be reverting this change

File: legend-engine-test-runner-service/src/test/java/org/finos/legend/engine/testable/service/TestServiceTestSuite.java [hunk 2]
```diff
@@ -2500,6 +2498,40 @@ public void testFailsWithKeysInSingleExec()
...
+//        Assert.assertTrue(inlineServiceStoreTestResults.get(0) instanceof MultiExecutionServiceTestResult);
+//        Assert.assertEquals("testModelStoreTestSuites::service::DocM2MService", inlineServiceStoreTestResults.get(0).testable);
+//        Assert.assertEquals("testSuite1", inlineServiceStoreTestResults.get(0).atomicTestId.testSuiteId);
+//        Assert.assertEquals("test1", inlineServiceStoreTestResults.get(0).atomicTestId.atomicTestId);
+//
+//        TestResult inlineUatTestResult = ((MultiExecutionServiceTestResult) inlineServiceStoreTestResults.get(0)).getKeyIndexedTestResults().get("UAT");
+//        Assert.assertTrue(inlineUatTestResult instanceof TestPassed);
+//        Assert.assertEquals("testModelStoreTestSuites::service::DocM2MService", inlineUatTestResult.testable);
+//        Assert.assertEquals("testSuite1", inlineUatTestResult.atomicTestId.testSuiteId);
+//        Assert.assertEquals("test1", inlineUatTestResult.atomicTestId.atomicTestId);
```
Comment (gs-ssh16) [inline legend-engine-test-runner-service/src/test/java/org/finos/legend/engine/testable/service/TestServiceTestSuite.java:2517]: Commented code

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/handlers/Handlers.java [hunk 3]
```diff
@@ -855,6 +862,7 @@ private void registerStrings()
                 m(h("meta::pure::functions::string::joinStrings_String_MANY__String_1__String_1__String_1__String_1_", true, ps -> res("String", "one"), ps -> ps.size() == 4)),
                 m(h("meta::pure::functions::string::joinStrings_String_MANY__String_1_", false, ps -> res("String", "one"), ps -> true))));
         register("meta::pure::functions::string::lastIndexOf_String_1__String_1__Integer_1_", false, ps -> res("Integer", "one"));
+        register("meta::pure::functions::string::lastIndexOf_String_1__String_1__Integer_1_", false, ps -> res("Integer", "one"));
```
Comment (gs-ssh16) [inline legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/handlers/Handlers.java:865]: Why duplicate?

File: legend-engine-language-pure-dsl-service/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/ServiceParserGrammar.g4 [hunk 4]
```diff
@@ -230,4 +231,28 @@ testSingleValueParam:                        primitiveValue
 ;
 
 testParam:                              testListValueParam | testSingleValueParam
+;
+
+// ----------------------------------- EXECUTION_ENVIRONMENT ------------------------------------------------------
+execEnvs:                               EXEC_ENV qualifiedName
```
Comment (gs-ssh16) [inline legend-engine-language-pure-dsl-service/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/ServiceParserGrammar.g4:237]: Its only one exec env right? why plural execEnvs

File: legend-engine-language-pure-dsl-service/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/ServiceParserGrammar.g4 [hunk 5]
```diff
@@ -230,4 +231,28 @@ testSingleValueParam:                        primitiveValue
...
+                                            BRACE_OPEN
+                                                execEnvType
+                                                executions
+                                            BRACE_CLOSE
+;
+execEnvType:                            EXEC_ENV_TYPE COLON execEnvTypes SEMI_COLON
+;
+execEnvTypes:                           SERVICE_SINGLE| SERVICE_MULTI
```
Comment (gs-ssh16) [inline legend-engine-language-pure-dsl-service/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/ServiceParserGrammar.g4:245]: This is kind of confusing - Multi exec service needs SingleExecEnv. May be we need a different name?

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mapping/mappingExtension.pure [hunk 6]
```diff
@@ -317,6 +317,16 @@ function  meta::pure::mapping::from(tds:TabularDataSet[1], mapping:Mapping[1], r
    $tds;
 }
 
+function  meta::pure::mapping::from<T|m>(t:T[m], executionParameters: meta::legend::service::metamodel::SingleExecutionParameters[1]):T[m]
+{
+   $t
+}
+
+function  meta::pure::mapping::from(tds:TabularDataSet[1], executionParameters: meta::legend::service::metamodel::SingleExecutionParameters[1]):TabularDataSet[1]
```
Comment (gs-ssh16) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mapping/mappingExtension.pure:325]: Do we need this function?

The above should handle both

File: legend-engine-test-runner-service/src/test/java/org/finos/legend/engine/testable/service/TestServiceTestSuite.java [hunk 7]
```diff
@@ -2500,6 +2498,40 @@ public void testFailsWithKeysInSingleExec()
...
+
+        // execution environment mentioned with reference in the service
+        List<TestResult> inlineServiceStoreTestResults = executeServiceTest("testable/m2m/","legend-testable-m2m-service-model.pure","legend-testable-m2m-inline-multiExec-embeddedParam.pure", "testModelStoreTestSuites::service::DocM2MService");
```
Comment (gs-ssh16) [inline legend-engine-test-runner-service/src/test/java/org/finos/legend/engine/testable/service/TestServiceTestSuite.java:2520]: Is the below file even used in tests?

File: legend-engine-language-pure-dsl-service/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/ServiceParserGrammar.g4 (hunk 5 above)
Comment (gs-gunjan) [reply legend-engine-language-pure-dsl-service/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/ServiceParserGrammar.g4:245]: well lexer doesnt alloy adding two different names with the same value

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mapping/mappingExtension.pure (hunk 6 above)
Comment (gs-gunjan) [reply legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mapping/mappingExtension.pure:325]: with from function I saw this pattern, so added it, but if not required would delete

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/handlers/Handlers.java (hunk 3 above)
Comment (gs-gunjan) [reply legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/handlers/Handlers.java:865]: merge issues

File: legend-engine-test-runner-service/src/test/resources/testable/m2m/legend-testable-m2m-inline-multiExec-embeddedParam.pure [hunk 8]
```diff
@@ -0,0 +1,75 @@
+###Service
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply legend-engine-test-runner-service/src/test/resources/testable/m2m/legend-testable-m2m-inline-multiExec-embeddedParam.pure:1]: we have not been adding this to any pf the pure resources file

File: legend-engine-test-runner-service/src/test/resources/testable/m2m/legend-testable-m2m-inline-multi-execution-service.pure [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  owners:
+  [
+    'dummy',
+    'dummy1'
+  ];
+  documentation: 'Service to test refiner flow';
+  autoActivateUpdates: true;
+  execution: Multi
+  {
+    query: env:String[1]|testModelStoreTestSuites::model::Doc.all()->graphFetchChecked(#{testModelStoreTestSuites::model::Doc{firm_tbl{addressId,firmId,legalName,ceoId},person_tbl{addressId,age,firmId,firstName,id,lastName}}}#)->serialize(#{testModelStoreTestSuites::model::Doc{firm_tbl{addressId,firmId,legalName,ceoId},person_tbl{addressId,age,firmId,firstName,id,lastName}}}#)->from($env,[]);
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply legend-engine-test-runner-service/src/test/resources/testable/m2m/legend-testable-m2m-inline-multi-execution-service.pure]: yep, this file was not being used at all, deleting it

File: legend-engine-language-pure-dsl-service/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/ServiceParserGrammar.g4 (hunk 5 above)
Comment (gs-gunjan) [reply legend-engine-language-pure-dsl-service/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/ServiceParserGrammar.g4]: renaming it SINGLE_KEYWORD and MULTI_KEYWORD

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/handlers/Handlers.java [hunk 10]
```diff
@@ -597,6 +597,11 @@ public Handlers(PureModel pureModel)
         register("meta::json::schema::mapSchema_String_1__Type_1__DiscriminatorMapping_1_", false, ps -> res("meta::json::schema::DiscriminatorMapping", "one"));
         register("meta::json::schema::discriminateOneOf_Any_1__Any_1__Type_MANY__DiscriminatorMapping_MANY__Boolean_1_", false, ps -> res("Boolean", "one"));
 
+        // getter for execution parameters from execution environment
+        register(m(
+                m(h("meta::legend::service::get_ExecutionEnvironmentInstance_1__String_$0_1$__SingleExecutionParameters_1_", false, ps -> res("meta::legend::service::metamodel::SingleExecutionParameters", "one"), ps -> ps.size() == 2)),
```
Comment (gs-ssh16) [inline legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/handlers/Handlers.java:602]: Why is masterKey optional?

Comment (gs-gunjan) [reply legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/handlers/Handlers.java]: changed it to non be optional for both this and single execution parameters

**Review by Yasirmod17 — APPROVED** [review]

File: legend-engine-test-runner-service/src/main/java/org/finos/legend/engine/testable/service/extension/ServiceTestRunner.java [hunk 11]
```diff
@@ -164,28 +137,115 @@ else if (service.execution instanceof PureSingleExecution)
...
+            for (Test test : suite.tests)
+            {
+                if (testIds.contains(test.id))
+                {
+                    List<String> keys;
+                    if (((ServiceTest) test).parameters != null)
+                    {
+                        String keyName = ((ServiceTest) test).parameters.stream()
+                                .filter(parameterValue -> parameterValue.name.equals(finalExeckey))
+                                .map(param -> param.value.accept(new PrimitiveValueSpecificationToObjectVisitor())).findFirst().get().toString();
```
Comment (rafaelbey) [inline legend-engine-test-runner-service/src/main/java/org/finos/legend/engine/testable/service/extension/ServiceTestRunner.java:180]: This line assume the optional always has a value, while integration testing failed with empty value.

**Pull request merged**